File: app/src/main/java/me/devsaki/hentoid/fragments/reader/ReaderNavigation.java
package me.devsaki.hentoid.fragments.reader;

import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.annimon.stream.Stream;
import com.google.android.material.slider.Slider;

import java.util.List;
import java.util.Locale;

import javax.annotation.Nonnull;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.databinding.FragmentReaderPagerBinding;
import me.devsaki.hentoid.databinding.IncludeReaderControlsOverlayBinding;
import me.devsaki.hentoid.ui.InputDialog;
import me.devsaki.hentoid.util.Helper;
import me.devsaki.hentoid.util.Preferences;
import me.devsaki.hentoid.util.ToastHelper;

class ReaderNavigation {

    // == UI
    private FragmentReaderPagerBinding superBinding;
    private IncludeReaderControlsOverlayBinding binding;
    // Bottom bar controls (proxies for left or right position, depending on current reading direction)
    private TextView pageCurrentNumber;
    private TextView pageMaxNumber;
    private View prevFunctionalButton;
    private View nextFunctionalButton;

    // == VARS
    private final Pager pager;
    private List<ImageFile> images;
    private List<Chapter> chapters;
    private Chapter currentChapter;
    private int maxPageNumber; // Relative (chapter-scale) max page number

    boolean isContentFirst = false;
    boolean isContentLast = false;


    public ReaderNavigation(Pager pager, FragmentReaderPagerBinding inBinding) {
        superBinding = inBinding;
        binding = inBinding.controlsOverlay;
        this.pager = pager;

        binding.pageSlider.addOnChangeListener((slider1, value, fromUser) -> {
            if (fromUser) {
                int offset = 0;
                if (Preferences.isReaderChapteredNavigation()) {
                    Chapter currentChapter = getCurrentChapter();
                    if (currentChapter != null) {
                        List<ImageFile> chapImgs = currentChapter.getReadableImageFiles();
                        if (!chapImgs.isEmpty()) offset = chapImgs.get(0).getOrder() - 1;
                    }
                }

                pager.seekToPosition(Math.max(0, offset + (int) value));
            }
        });

        binding.pageSlider.addOnSliderTouchListener(new Slider.OnSliderTouchListener() {
            @Override
            public void onStartTrackingTouch(@NonNull Slider slider) {
                binding.imagePreviewLeft.setVisibility(View.VISIBLE);
                binding.imagePreviewCenter.setVisibility(View.VISIBLE);
                binding.imagePreviewRight.setVisibility(View.VISIBLE);
                superBinding.recyclerView.setVisibility(View.INVISIBLE);
            }

            @Override
            public void onStopTrackingTouch(@NonNull Slider slider) {
                binding.imagePreviewLeft.setVisibility(View.INVISIBLE);
                binding.imagePreviewCenter.setVisibility(View.INVISIBLE);
                binding.imagePreviewRight.setVisibility(View.INVISIBLE);
                superBinding.recyclerView.setVisibility(View.VISIBLE);
            }
        });
    }

    void clear() {
        binding = null;
        superBinding = null;
    }

    /**
     * Update the visibility of "next/previous book" buttons
     *
     * @param content Current book
     */
    void onContentChanged(@Nonnull Content content) {
        int direction = Preferences.getContentDirection(content.getBookPreferences());
        nextFunctionalButton = (Preferences.Constant.VIEWER_DIRECTION_LTR == direction) ? binding.nextBookBtn : binding.prevBookBtn;
        prevFunctionalButton = (Preferences.Constant.VIEWER_DIRECTION_LTR == direction) ? binding.prevBookBtn : binding.nextBookBtn;

        isContentFirst = content.isFirst();
        isContentLast = content.isLast();
    }

    void onImagesChanged(@NonNull List<ImageFile> images) {
        this.images = images;
        chapters = Stream.of(images).map(ImageFile::getLinkedChapter).withoutNulls().sortBy(Chapter::getOrder).distinct().toList();

        // Can't access the gallery when there's no page to display
        if (images.size() > 0) binding.galleryBtn.setVisibility(View.VISIBLE);
        else binding.galleryBtn.setVisibility(View.GONE);

        maxPageNumber = (int) Stream.of(images).filter(ImageFile::isReadable).count();
    }

    private void onChapterChanged(@NonNull Chapter chapter) {
        ToastHelper.toast(chapter.getName());
        updateNextPrevButtonsChapter(chapter);
    }

    private void updateNextPrevButtonsChapter(Chapter chapter) {
        int chapterIndex = getChapterIndex(chapter);
        boolean isChapterFirst = 0 == chapterIndex;
        boolean isChapterLast = chapters.size() - 1 == chapterIndex;
        prevFunctionalButton.setVisibility(isChapterFirst ? View.INVISIBLE : View.VISIBLE);
        nextFunctionalButton.setVisibility(isChapterLast ? View.INVISIBLE : View.VISIBLE);
    }

    void setDirection(int direction) {
        if (Preferences.Constant.VIEWER_DIRECTION_LTR == direction) {
            pageCurrentNumber = binding.pagerLeftTxt;
            pageMaxNumber = binding.pagerRightTxt;
            binding.pageSlider.setRotationY(0);
            binding.prevBookBtn.setOnClickListener(v -> previousFunctional());
            binding.nextBookBtn.setOnClickListener(v -> nextFunctional());
        } else if (Preferences.Constant.VIEWER_DIRECTION_RTL == direction) {
            pageCurrentNumber = binding.pagerRightTxt;
            pageMaxNumber = binding.pagerLeftTxt;
            binding.pageSlider.setRotationY(180);
            binding.prevBookBtn.setOnClickListener(v -> nextFunctional());
            binding.nextBookBtn.setOnClickListener(v -> previousFunctional());
        }

        pageMaxNumber.setOnClickListener(null);
        pageCurrentNumber.setOnClickListener(v -> InputDialog.invokeNumberInputDialog(binding.pageSlider.getContext(), R.string.goto_page, pager::goToPage));
    }

    /**
     * Update the display of page position controls (text and bar)
     */
    void updatePageControls() {
        ImageFile img = pager.getCurrentImg();
        if (null == img) return;

        int pageNum = img.getOrder();
        int pageOffset = 0;
        if (Preferences.isReaderChapteredNavigation()) {
            Chapter newChap = getCurrentChapter();
            if (newChap != null) {
                if (null != currentChapter && newChap.uniqueHash() != currentChapter.uniqueHash())
                    onChapterChanged(newChap);
                List<ImageFile> chapImgs = newChap.getReadableImageFiles();
                // Caution : ImageFiles inside chapters don't have any displayed order
                // To get it, a mapping between image list and chapters has to be done
                if (!chapImgs.isEmpty()) pageOffset = chapImgs.get(0).getOrder();
                pageNum = pageNum - pageOffset + 1;
            }
            currentChapter = newChap;
        } else {
            currentChapter = null;
        }

        int maxPageNum = maxPageNumber;
        if (Preferences.isReaderChapteredNavigation()) {
            if (currentChapter != null) maxPageNum = currentChapter.getReadableImageFiles().size();
        }

        pageCurrentNumber.setText(String.format(Locale.ENGLISH, "%d", pageNum));
        pageMaxNumber.setText(String.format(Locale.ENGLISH, "%d", maxPageNum));
        superBinding.viewerPagenumberText.setText(String.format(Locale.ENGLISH, "%d / %d", pageNum, maxPageNum));

        // Only update slider when user isn't skimming _with_ the slider
        if (binding.imagePreviewCenter.getVisibility() != View.VISIBLE) {
            int sliderMaxPos = Math.max(1, maxPageNum - 1);
            // Next line to avoid setting a max position inferior to current position
            binding.pageSlider.setValue(Helper.coerceIn(binding.pageSlider.getValue(), 0, sliderMaxPos));
            binding.pageSlider.setValueTo(sliderMaxPos);

            int imageIndex = getCurrentImageIndex();
            if (imageIndex > -1)
                binding.pageSlider.setValue(Helper.coerceIn(imageIndex, 0, sliderMaxPos));
        }

        if (!Preferences.isReaderChapteredNavigation() || null == chapters || chapters.isEmpty()) {
            prevFunctionalButton.setVisibility(isContentFirst ? View.INVISIBLE : View.VISIBLE);
            nextFunctionalButton.setVisibility(isContentLast ? View.INVISIBLE : View.VISIBLE);
        } else updateNextPrevButtonsChapter(currentChapter);
    }

    void previousFunctional() {
        if (!Preferences.isReaderChapteredNavigation()) pager.previousBook();
        else if (!previousChapter()) pager.previousBook();
    }

    void nextFunctional() {
        if (!Preferences.isReaderChapteredNavigation()) pager.nextBook();
        else if (!nextChapter()) pager.nextBook();
    }

    private boolean previousChapter() {
        int currentChIndex = getCurrentChapterIndex();
        if (currentChIndex > 0) {
            Chapter selectedChapter = chapters.get(currentChIndex - 1);
            List<ImageFile> chImgs = selectedChapter.getReadableImageFiles();
            if (chImgs.isEmpty()) return false;
            pager.goToPage(chImgs.get(0).getOrder());
            return true;
        }
        return false;
    }

    private boolean nextChapter() {
        int currentChIndex = getCurrentChapterIndex();
        if (currentChIndex < chapters.size() - 1) {
            Chapter selectedChapter = chapters.get(currentChIndex + 1);
            List<ImageFile> chImgs = selectedChapter.getReadableImageFiles();
            if (chImgs.isEmpty()) return false;
            pager.goToPage(chImgs.get(0).getOrder());
            return true;
        }
        return false;
    }

    private int getChapterIndex(Chapter ch) {
        if (null == ch || null == chapters || chapters.isEmpty()) return -1;

        for (int i = 0; i < chapters.size(); i++)
            if (chapters.get(i).getId() == ch.getId()) return i;

        return -1;
    }

    private int getCurrentChapterIndex() {
        return getChapterIndex(getCurrentChapter());
    }

    private Chapter getCurrentChapter() {
        ImageFile startFrom = pager.getCurrentImg();
        if (null == startFrom) return null;

        return startFrom.getLinkedChapter();
    }

    private int getCurrentImageIndex() {
        ImageFile currentImg = pager.getCurrentImg();
        Chapter currentChapter = getCurrentChapter();
        // Absolute index
        if (!Preferences.isReaderChapteredNavigation() || null == currentChapter) {
            return indexAmong(currentImg, images);
        } else { // Relative to current chapter
            return indexAmong(currentImg, currentChapter.getReadableImageFiles());
        }
    }

    private int indexAmong(ImageFile img, List<ImageFile> imgs) {
        if (img != null && imgs != null) {
            for (int i = 0; i < imgs.size(); i++)
                if (imgs.get(i).uniqueHash() == img.uniqueHash()) return i;
        }
        return -1;
    }

    interface Pager {
        void goToPage(int absPageNum);

        void seekToPosition(int absIndex);

        void nextBook();

        void previousBook();

        ImageFile getCurrentImg();
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/reader/ReaderNavigation.kt
package me.devsaki.hentoid.fragments.reader

import android.view.View
import android.widget.TextView
import androidx.core.view.isVisible
import com.google.android.material.slider.Slider
import me.devsaki.hentoid.R
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.databinding.FragmentReaderPagerBinding
import me.devsaki.hentoid.databinding.IncludeReaderControlsOverlayBinding
import me.devsaki.hentoid.ui.invokeNumberInputDialog
import me.devsaki.hentoid.util.Settings
import me.devsaki.hentoid.util.Settings.Value.VIEWER_DIRECTION_LTR
import me.devsaki.hentoid.util.Settings.Value.VIEWER_DIRECTION_RTL
import me.devsaki.hentoid.util.coerceIn
import me.devsaki.hentoid.util.toast
import java.util.Locale

class ReaderNavigation(private val pager: Pager, inBinding: FragmentReaderPagerBinding) {
    // == UI
    private var superBinding: FragmentReaderPagerBinding? = null
    private var binding: IncludeReaderControlsOverlayBinding? = null

    // Bottom bar controls (proxies for left or right position, depending on current reading direction)
    private var pageCurrentNumber: TextView? = null
    private var pageMaxNumber: TextView? = null
    private var prevFunctionalButton: View? = null
    private var nextFunctionalButton: View? = null

    // == VARS
    private var images: List<ImageFile>? = null
    private var chapters: List<Chapter>? = null
    private var currentChapter: Chapter? = null

    // Relative (chapter-scale) max page number
    private var maxPageNumber = 0
    private var isContentDynamic = false

    private var isContentFirst = false
    private var isContentLast = false


    init {
        superBinding = inBinding
        inBinding.controlsOverlay.let {
            binding = it
            it.pageSlider.addOnChangeListener { _, value, fromUser ->
                if (fromUser) {
                    var offset = 0
                    if (Settings.isReaderChapteredNavigation) {
                        getCurrentChapter()?.apply {
                            val chapImgs = readableImageFiles
                            if (chapImgs.isNotEmpty()) offset =
                                pager.indexFromPageNum(chapImgs[0].order)
                        }
                    }
                    pager.seekToIndex(0.coerceAtLeast(offset + value.toInt()))
                }
            }
            it.pageSlider.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
                override fun onStartTrackingTouch(slider: Slider) {
                    it.imagePreviewLeft.visibility = View.VISIBLE
                    it.imagePreviewCenter.visibility = View.VISIBLE
                    it.imagePreviewRight.visibility = View.VISIBLE
                    superBinding?.recyclerView?.visibility = View.INVISIBLE
                }

                override fun onStopTrackingTouch(slider: Slider) {
                    it.imagePreviewLeft.visibility = View.INVISIBLE
                    it.imagePreviewCenter.visibility = View.INVISIBLE
                    it.imagePreviewRight.visibility = View.INVISIBLE
                    superBinding?.recyclerView?.visibility = View.VISIBLE
                }
            })
        }
    }

    fun clear() {
        binding = null
        superBinding = null
    }

    /**
     * Update the visibility of "next/previous book" buttons
     *
     * @param content Current book
     */
    fun onContentChanged(content: Content) {
        val direction = Settings.getContentDirection(content.bookPreferences)
        nextFunctionalButton =
            if (VIEWER_DIRECTION_LTR == direction) binding?.nextBookBtn else binding?.prevBookBtn
        prevFunctionalButton =
            if (VIEWER_DIRECTION_LTR == direction) binding?.prevBookBtn else binding?.nextBookBtn
        isContentFirst = content.isFirst
        isContentLast = content.isLast
        isContentDynamic = content.isDynamic
    }

    fun onImagesChanged(images: List<ImageFile>) {
        this.images = images
        chapters = images.mapNotNull { obj: ImageFile -> obj.linkedChapter }
            .sortedBy { obj: Chapter -> obj.order }.distinct().toList()

        // Can't access the gallery when there's no page to display
        binding?.apply {
            if (images.isNotEmpty()) galleryBtn.visibility = View.VISIBLE
            else galleryBtn.visibility = View.GONE
        }
        maxPageNumber = images.count { obj: ImageFile -> obj.isReadable }
    }

    private fun onChapterChanged(chapter: Chapter) {
        pageCurrentNumber?.context?.toast(chapter.name)
        updateNextPrevButtonsChapter(chapter)
    }

    private fun updateNextPrevButtonsChapter(chapter: Chapter?) {
        val chapterIndex = getChapterIndex(chapter)
        val isChapterFirst = 0 == chapterIndex
        val isChapterLast = chapters!!.size - 1 == chapterIndex
        prevFunctionalButton?.visibility = if (isChapterFirst) View.INVISIBLE else View.VISIBLE
        nextFunctionalButton?.visibility = if (isChapterLast) View.INVISIBLE else View.VISIBLE
    }

    fun setDirection(direction: Int) {
        binding?.apply {
            if (VIEWER_DIRECTION_LTR == direction) {
                pageCurrentNumber = pagerLeftTxt
                pageMaxNumber = pagerRightTxt
                pageSlider.rotationY = 0f
                prevBookBtn.setOnClickListener { previousContainer() }
                nextBookBtn.setOnClickListener { nextContainer() }
            } else if (VIEWER_DIRECTION_RTL == direction) {
                pageCurrentNumber = pagerRightTxt
                pageMaxNumber = pagerLeftTxt
                pageSlider.rotationY = 180f
                prevBookBtn.setOnClickListener { nextContainer() }
                nextBookBtn.setOnClickListener { previousContainer() }
            }
            pageMaxNumber?.setOnClickListener(null)
            pageCurrentNumber?.setOnClickListener {
                invokeNumberInputDialog(
                    pageSlider.context,
                    R.string.goto_page
                ) { absPageNum: Int ->
                    pager.goToPage(absPageNum)
                }
            }
        }
    }

    /**
     * Update the display of page position controls (text and bar)
     */
    fun updatePageControls() {
        val img = pager.currentImg ?: return
        var pageNum = img.displayOrder + 1
        var pageOffset = 0
        if (Settings.isReaderChapteredNavigation && !isContentDynamic) {
            val newChap = getCurrentChapter()
            if (newChap != null) {
                if (null != currentChapter && newChap.uniqueHash() != currentChapter!!.uniqueHash())
                    onChapterChanged(newChap)
                val chapImgs = newChap.readableImageFiles
                // Caution : ImageFiles inside chapters don't have any displayed order
                // To get it, a mapping between image list and chapters has to be done
                if (chapImgs.isNotEmpty()) pageOffset = chapImgs[0].order
                pageNum = pageNum - pageOffset + 1
            }
            currentChapter = newChap
        } else {
            currentChapter = null
        }
        var maxPageNum = maxPageNumber
        if (Settings.isReaderChapteredNavigation && currentChapter != null) {
            maxPageNum = currentChapter!!.readableImageFiles.size
        }
        pageCurrentNumber?.text = String.format(Locale.ENGLISH, "%d", pageNum)
        pageMaxNumber?.text = String.format(Locale.ENGLISH, "%d", maxPageNum)
        superBinding?.viewerPagenumberText?.text = String.format(
            Locale.ENGLISH,
            "%d / %d",
            pageNum,
            maxPageNum
        )

        // Only update slider when user isn't skimming _with_ the slider
        binding?.let {
            if (!it.imagePreviewCenter.isVisible) {
                val sliderMaxPos = 1.coerceAtLeast(maxPageNum - 1)
                // Next line to avoid setting a max position inferior to current position
                it.pageSlider.value = coerceIn(
                    it.pageSlider.value,
                    0f,
                    sliderMaxPos.toFloat()
                )
                it.pageSlider.valueTo = sliderMaxPos.toFloat()
                val imageIndex = getCurrentImageIndex()
                if (imageIndex > -1) it.pageSlider.value = coerceIn(
                    imageIndex.toFloat(),
                    0f,
                    sliderMaxPos.toFloat()
                )
            }
        }
        if (!Settings.isReaderChapteredNavigation || null == chapters || chapters!!.isEmpty()) {
            prevFunctionalButton?.visibility = if (isContentFirst) View.INVISIBLE else View.VISIBLE
            nextFunctionalButton?.visibility = if (isContentLast) View.INVISIBLE else View.VISIBLE
        } else updateNextPrevButtonsChapter(currentChapter)
    }

    fun previousContainer(): Boolean {
        return if (!Settings.isReaderChapteredNavigation) pager.previousBook()
        else if (!previousChapter()) pager.previousBook()
        else true
    }

    fun nextContainer(): Boolean {
        return if (!Settings.isReaderChapteredNavigation) pager.nextBook()
        else if (!nextChapter()) pager.nextBook()
        else true
    }

    private fun previousChapter(): Boolean {
        val currentChIndex = getCurrentChapterIndex()
        if (currentChIndex > 0) {
            chapters?.let {
                val selectedChapter = it[currentChIndex - 1]
                val chImgs = selectedChapter.readableImageFiles
                if (chImgs.isEmpty()) return false
                pager.goToPage(chImgs[0].order)
                return true
            }
        }
        return false
    }

    private fun nextChapter(): Boolean {
        val currentChIndex = getCurrentChapterIndex()
        chapters?.let {
            if (currentChIndex < it.size - 1) {
                val selectedChapter = it[currentChIndex + 1]
                val chImgs = selectedChapter.readableImageFiles
                if (chImgs.isEmpty()) return false
                pager.goToPage(chImgs[0].order)
                return true
            }
        }
        return false
    }

    private fun getChapterIndex(ch: Chapter?): Int {
        if (null == ch || null == chapters) return -1
        chapters?.let {
            if (it.isEmpty()) return -1
            for (i in it.indices) if (it[i].id == ch.id) return i
        }
        return -1
    }

    private fun getCurrentChapterIndex(): Int {
        return getChapterIndex(getCurrentChapter())
    }

    private fun getCurrentChapter(): Chapter? {
        val startFrom = pager.currentImg ?: return null
        return startFrom.linkedChapter
    }

    private fun getCurrentImageIndex(): Int {
        val currentImg = pager.currentImg
        val currentChapter = getCurrentChapter()
        // Absolute index
        return if (!Settings.isReaderChapteredNavigation || null == currentChapter) {
            indexAmong(currentImg, images)
        } else { // Relative to current chapter
            indexAmong(currentImg, currentChapter.readableImageFiles)
        }
    }

    private fun indexAmong(img: ImageFile?, imgs: List<ImageFile>?): Int {
        if (img != null && imgs != null) {
            for (i in imgs.indices) if (imgs[i].id == img.id) return i
        }
        return -1
    }

    interface Pager {
        fun goToPage(absPageNum: Int)
        fun seekToIndex(absIndex: Int)
        fun indexFromPageNum(pageNum: Int): Int
        fun nextBook(): Boolean
        fun previousBook(): Boolean
        val currentImg: ImageFile?
    }
}